README.md
native-node-dns -- A replacement DNS stack for node.js
=================================================

# New Fork - native-node-dns

Francis Turner has forked the original node-dns repo to make this new
repo and npm package and the associated native-dns-cache 
and native-dns-packet repos/packages (as native-node-dns-cache and 
native-node-dns-packet)

Due to the way github works, please file issues in the original repo (tjfontaine/node-dns ) and tag me
until I can figure out how to make this a separate project

# Original README (with updates from incorporated pull reqs)

Installation
------------

```
npm install native-node-dns
```

Client
------

native-node-dns exports what should be a 1:1 mapping of the upstream node.js dns
module. That is to say if it's listed in the [docs](http://nodejs.org/docs/latest/api/dns.html)
it should behave similarly. If it doesn't please file an [issue](https://github.com/tjfontaine/node-dns/issues/new).

Request
-------

Beyond matching the upstream module, native-node-dns also provides a method for
customizing queries.

```javascript
var dns = require('native-node-dns');
var util = require('util');

var question = dns.Question({
  name: 'www.google.com',
  type: 'A',
});

var start = Date.now();

var req = dns.Request({
  question: question,
  server: { address: '8.8.8.8', port: 53, type: 'udp' },
  timeout: 1000,
});

req.on('timeout', function () {
  console.log('Timeout in making request');
});

req.on('message', function (err, answer) {
  answer.answer.forEach(function (a) {
    console.log(a.address);
  });
});

req.on('end', function () {
  var delta = (Date.now()) - start;
  console.log('Finished processing request: ' + delta.toString() + 'ms');
});

req.send();
```

Request creation takes an object with the following fields

 * `question` -- an instance of Question (required)
 * `server` -- defines the remote end point (required)
  - as an object it should be
    * `address` -- a string ip address (required)
    * `port` -- a number for the remote port (optional, default 53)
    * `type` -- a string indicating `udp` or `tcp` (optional, default `udp`)
You do not need to indicate ipv4 or ipv6, the backend will handle that
  - a string ip address
 * `additional` -- an array of Additional Resource Records (optional)
 * `timeout` -- a number in milliseconds indicating how long to wait for the
request to finish. (optional, default 4000)
 * `try_edns` -- a boolean indicating whether to use an `EDNSPacket` (optional)
 * `cache` -- can be false to disable caching, or implement the cache model, or
an instance of Cache but with a different store (optional, default
platform.cache)
 * `header` -- an object containing header flags (optional)

There are only two methods

 * `send` -- sends the actual request to the remote endpoint
 * `cancel` -- cancels the request and ignores any responses

Request emits the following events

 * `message` -- This is where you get a response, passes `(err, answer)` where
answer is an instance of `Packet`
 * `timeout` -- Fired when the timeout is reached
 * `cancelled` -- Fired if the request is cancelled
 * `end` -- Always fired after a request finished, regardless of disposition

Platform
--------

If you want to customize all `resolve` or `lookup`s with the replacement client
stack you can modify the platform settings accessible in the top level `platform`
object.

Methods:

 * `reload` -- Re-read system configuration files to populate name servers and
hosts

Properties:

 * `ready` -- Boolean whether requests are safe to transit, true after hosts
and name servers are filled
 * `watching` -- Boolean indicating if system configuration files are watched
for changes, default to false (currently can only be enabled on !win32)
 * `name_servers` -- An array of servers used for resolving queries against
  - Each entry is an object of `{ address: <string ip>, port: 53 }`
  - On win32 this is hard coded to be google dns until there's a sane way to get
the data
 * `search_path` -- An array of domains to try and append after a failed lookup
 * `attempts` -- The number of retries for a failed lookup/timeout (default: 5)
 * `timeout` -- The time each query is allowed to take before trying another
server. (in milliseconds, default: 5000 (5 seconds))
 * `edns` -- Whether to try and send edns queries first (default: false)
 * `cache` -- The system wide cache used by default for `lookup` and `resolve`,
set this to false to disable caching

Events:

 * `ready` -- Emitted after hosts and name servers have been loaded
 * `unready` -- Emitted when hosts and name servers configuration is being
reloaded.

Server
------

There is also a rudimentary server implementation

```javascript
var dns = require('native-node-dns');
var server = dns.createServer();

server.on('request', function (request, response) {
  //console.log(request)
  response.answer.push(dns.A({
    name: request.question[0].name,
    address: '127.0.0.1',
    ttl: 600,
  }));
  response.answer.push(dns.A({
    name: request.question[0].name,
    address: '127.0.0.2',
    ttl: 600,
  }));
  response.additional.push(dns.A({
    name: 'hostA.example.org',
    address: '127.0.0.3',
    ttl: 600,
  }));
  response.send();
});

server.on('error', function (err, buff, req, res) {
  console.log(err.stack);
});

server.serve(15353);
```

Server creation

 * `createServer` and `createUDPServer` -- both create a `UDP` based server,
they accept an optional object for configuration,
  - `{ dgram_type: 'udp4' }` is the default option, the other is `udp6`
 * `createTCPServer` -- creates a TCP based server

Server methods

 * `serve(port, [address])` -- specify which port and optional address to listen
on
 * `close()` -- stop the server/close sockets.

Server events

 * `listening` -- emitted when underlying socket is listening
 * `close` -- emitted when the underlying socket is closed
 * `request` -- emitted when a dns message is received, and the packet was
successfully unpacked, passes `(request, response)`
  - Both `request` and `response` are instances of `Packet` when you're finished
creating the response, you merely need to call `.send()` and the packet will
DoTheRightThing
 * `error` -- emitted when unable to properly unpack the packet, passed `(err, msg, response)`
 * `socketError` -- remap of the underlying socket for the server, passes `(err, socket)`

Packet
------

Properties:

 * `header`
  - `id` -- request id
  - `qdcount` -- the number of questions (inferred from array size)
  - `ancount` -- the number of questions (inferred from array size)
  - `nscount` -- the number of questions (inferred from array size)
  - `arcount` -- the number of questions (inferred from array size)
  - `qr` -- is a query response
  - `opcode`
  - `aa` -- Authoritative Answer
  - `tc` -- Truncation bit
  - `rd` -- Recursion Desired
  - `ra` -- Recursion Available
  - `res1` -- Reserved field
  - `res2` -- Reserved field
  - `res3` -- Reserved field
  - `rcode` -- Response Code (see `consts.NAME_TO_RCODE`)
 * `question` -- array of `Question`s
 * `answer` -- array of `ResourceRecord`s
 * `authority` -- array of `ResourceRecord`s
 * `additional` -- array of `ResourceRecord`s

Methods:

 * `send()` -- Handles sending the packet to the right end point

Question
--------

A `Question` is instantiated by passing an object like:

 * `name` -- i.e. 'www.google.com' (required)
 * `type` -- Either the string representation of the record type, or the integer
value, see `consts.NAME_TO_QTYPE` (default: 'A')
 * `class` -- The class of service, default to 1 meaning internet

ResourceRecord
--------------

ResourceRecords are what populate `answer`, `authority`, and `additional`.
This is a generic type, and each derived type inherits the following properties:

 * `name` -- The name of the resource
 * `type` -- The numerical representation of the resource record type
 * `class` -- The numerical representation of the class of service (usually 1 for internet)
 * `ttl` -- The Time To Live for the record, in seconds

Available Types:

 * `SOA`
  - `primary` -- string
  - `admin` -- string
  - `serial` -- number
  - `refresh` -- number
  - `retry` -- number
  - `expiration` -- number
  - `minimum` -- number
 * `A` and `AAAA`
  - `address` -- string
 * `MX`
  - `priority` -- number
  - `exchange` -- string
 * `TXT`
  - `data` -- array of strings
 * `SRV`
  - `priority` -- number
  - `weight` -- number
  - `port` -- number
  - `target` -- string
 * `NS`
  - `data` -- string
 * `CNAME`
  - `data` -- string
 * `PTR`
  - `data` -- string
 * `NAPTR`
  - `order` -- number
  - `preference` -- number
  - `flags` -- string
  - `service` -- string
  - `regexp` -- string
  - `replacement` -- string


ChangeLog
0.7.6: 2021-05-21 Francis Turner <francis.turner@gmail.com>
  * Update package.json to refer to correct github repo etc.
0.7.5: 2021-05-21 Francis Turner <francis.turner@gmail.com>
  * Update requires to use native-node-dns-packet etc.
0.7.4: 2021-05-21 Francis Turner <francis.turner@gmail.com>
  * Update requires to use native-node-dns-packet etc.
0.7.3: 2021-05-21 Francis Turner <francis.turner@gmail.com>
  * Fix package version issue
0.7.2: 2021-05-21 Francis Turner <francis.turner@gmail.com>
  * merge 3 outstanding PRs from original project (optional headers and
    additional records in request, test fixes)
  * change npm package to native-node-dns and use the similarly renamed
    native-node-dns-packet and native-node-dns-cache
0.7.1: 2021-05-20 Francis Turner <francis.turner@gmail.com>
  * change new Buffer() to new Buffer.alloc()
  * change TXT test to use a different domain the google.com so that
  * the harness works as expected

0.7.0: 2014-08-05 Greg Slepak <contact@taoeffect.com>
  * `native-dns-packet` 0.0.4 -> 0.1.1
  * `native-dns-cache`  0.0.1 -> 0.0.2
  * `ipaddr.js`         0.1.1 -> 0.1.3
  * Updated authors in `package.json`
  * Fixed specification of dependency versions in `package.json`
  * See changes from `native-dns-packet` for what's new
     + TLDR: many bug fixes, performance improvements.
             better EDNS support, TLSA support, fixed TXT.
  * Merged `tjfontaine/master` (adds LICENSE file)

2013-03-23  Timothy J Fontaine  <tjfontaine@gmail.com>
  * Add SPF record type, same as TXT
  * Emit platform ready on windows (Oleg Elifantiev)
  * Split packet parsing into its own library
  * Split caching into its own library

2013-03-03  Timothy J Fontaine  <tjfontaine@gmail.com>
  * Assert missing fields
  * Truncate integer fields

2013-02-14  Timothy J Fontaine  <tjfontaine@gmail.com>
  * Bump to v0.4.1
  * Disable caching for now
  * Fix AAAA packing of records

2013-01-09  Timothy J Fontaine  <tjfontaine@gmail.com>
  * Bump to v0.4.0
  * Fix TCPServer
  * Split dependencies out of package
  * Refactor Cache and MemoryStore
  * Export Lookup function

2012-12-23  Timothy J Fontaine  <tjfontaine@gmail.com>
  * Bump to v0.3.4
  * Update buffercursor to 0.0.5
  * Allow creation of NAPTR records

2012-12-07  Timothy J Fontaine  <tjfontaine@gmail.com>
  * Bump to v0.3.3
  * Make cache layer more generic

2012-12-06  Timothy J Fontaine  <tjfontaine@gmail.com>
  * Bump to v0.3.2
  * Revamp benchmark script
  * Properly set EDNS0 options
  * Don't overrite `type` and `class` fields of a record

2012-09-25  Timothy J Fontaine  <tjfontaine@gmail.com>
  * Bump to v0.3.1
  * Fix win32 to set the actual hostsfiel

2012-09-17  Timothy J Fontaine  <tjfontaine@gmail.com>
  * Bump to v0.3.0
  * Add more tests and benchmark
  * More compatible with c-ares
  * Add rudimentary in-memory cache
  * Add NAPTR record type
  * On > 0.8 make sockets be unref for faster destructing

2012-07-22  Timothy J Fontaine  <tjfontaine@gmail.com>
  * Bump to v0.2.0
  * Rewrite outbound queue to be simpler

2012-07-08  Timothy J Fontaine  <tjfontaine@gmail.com>
  * Bump to v0.1.0
  * State machine to encode and decode packets

2012-03-20  Timothy J Fontaine  <tjfontaine@gmail.com>
  * Bump to v0.0.7
  * Ignore requests/responses less than the minimal size of a query [server, client]
  * Remove clone dependency [module]
  * Embed ipaddr.js [module]
  * Fix label parsing [client, server]
  * Long query names should result in ENOTFOUND [client]
  * resolve(..., 'PTR') should use reverse, reverse should check hosts [client]
  * Autopromote should work for eDNS packets as well [client, server]
  * Fix server socket assignment [server]
  * toString to various types to produce dig like output [client, server]
  * Conosolidate various sources files into logical groups [module]
  * Lint to match upstream's style [module]

2012-03-06  Timothy J Fontaine  <tjfontaine@gmail.com>
  * Bump to v0.0.6
  * Split TCP/UDP Server, expose as createTCPServer/createUDPServer [server, module]
  * Request server may just be an string of an ip address [client]
  * Question type may be a string of record type [client]
  * Don't watch for resolve/hosts changes by default [client]
  * Add mechanism to forcibly reload the platform [client]
  * Add unready event that is fired when reinitializing platform [client]
  * Expose platform [module]

2012-03-01  Timothy J Fontaine  <tjfontaine@gmail.com>
  * Bump to v0.0.5
  * Add test suite [module] (issue #12)
  * Add qtypeToName/nameToQtype for convenience [module]
  * Add Request mechanism to cusotmize queries [module]
  * Add field name, buffer, and position to [un]pack errors [module]
  * Add PendingRequests.autopromote to always promote ResourceRecords [module]
  * Fix Platform.search_path it should be an array [client]

2012-02-25  Timothy J Fontaine  <tjfontaine@gmail.com>
  * Bump to v0.0.4
  * Add clone dependency (it's used in client) [module]
  * Don't instantiate fields for every question [client, server] (issue #8)
  * Allocate only the size Buffer needed for Packet [client, server]

2012-02-24  Timothy J Fontaine  <tjfontaine@gmail.com>
  * Bump to v0.0.3
  * Considerable performance increase [client, server] (issue #8)
  * Label compression is case sensitive [client, server]
  * Remove dependency on clone and pystruct [client, server]
  * Consolidate TCP logic [client, server] (issue #3)

2012-02-14  Timothy J Fontaine  <tjfontaine@gmail.com>
  * Bump to v0.0.2
  * Add rudimentary TCP support [client, server]
  * Rate limit queries to name servers [client]
  * Single buffer creation for message packing [client, server]
  * Label compression [client, server]
  * Expose consts for use by servers [module]
  * Specify specific server for querying against [client]
  * Remove dependency on bufferjs [module]


test/client.js
/*
The follwoing code adapted from node's test-dns.js license is as follows
*/

// Copyright Joyent, Inc. and other Node contributors.
//
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to permit
// persons to whom the Software is furnished to do so, subject to the
// following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
// USE OR OTHER DEALINGS IN THE SOFTWARE.

var dns = require('../dns'),
    net = require('net'),
    isIP = net.isIP,
    isIPv4 = net.isIPv4,
    isIPv6 = net.isIPv6;

var platform = require('../lib/platform');

platform.cache = false;

function checkWrap(test, req) {
  test.ok(typeof req === 'object');
}

var REMOTE_DNS = require('../test_utils').REMOTE_DNS;
var REMOTE_PORT = require('../test_utils').REMOTE_PORT;

var fixed = false;

var fixupDns = function () {
  platform.name_servers = [{
    address: REMOTE_DNS,
    port: REMOTE_PORT,
  }];

  platform.search_path = [];

  fixed = true;
};

platform.on('ready', fixupDns);

if (platform.ready)
  fixupDns();

exports.setUp = function (cb) {
  var checkReady = function() {
    if (fixed) {
      cb();
    } else {
      setImmediate(checkReady);
    }
  }
  checkReady();
};

exports.resolve4 = function (test) {
  var req = dns.resolve4('www.google.com', function(err, ips) {
    test.ifError(err);

    test.ok(ips);
    test.ok(ips.length > 0);

    for (var i = 0; i < ips.length; i++) {
      test.ok(isIPv4(ips[i]));
    }

    test.done();
  });

  checkWrap(test, req);
};


exports.resolve6 = function (test) {
  var req = dns.resolve6('ipv6.google.com', function(err, ips) {
    test.ifError(err);

    test.ok(ips);
    test.ok(ips.length > 0);

    for (var i = 0; i < ips.length; i++) {
      test.ok(isIPv6(ips[i]));
    }

    test.done();
  });

  checkWrap(test, req);
};


exports.reverse_ipv4 = function (test) {
  var req = dns.reverse('8.8.8.8', function(err, domains) {
    test.ifError(err);

    test.ok(domains);
    test.ok(domains.length > 0);

    for (var i = 0; i < domains.length; i++) {
      test.ok(domains[i]);
      test.ok(typeof domains[i] === 'string');
    }

    test.done();
  });

  checkWrap(test, req);
};


exports.reverse_ipv6 = function (test) {
  var req = dns.reverse('2001:4860:4860::8888', function(err, domains) {
    test.ifError(err);

    test.ok(domains);
    test.ok(domains.length > 0);

    for (var i = 0; i < domains.length; i++) {
      test.ok(domains[i]);
      test.ok(typeof domains[i] === 'string');
    }

    test.done();
  });

  checkWrap(test, req);
};


exports.reverse_bogus = function (test) {
  var error;

  try {
    var req = dns.reverse('bogus ip', function() {
      test.ok(false);
    });
  } catch (e) {
    error = e;
  }

  test.ok(error instanceof Error);
  test.strictEqual(error.errno, 'ENOTIMP');

  test.done();
};


exports.resolveMx = function (test) {
  var req = dns.resolveMx('gmail.com', function(err, result) {
    test.ifError(err);

    test.ok(result);
    test.ok(result.length > 0);

    for (var i = 0; i < result.length; i++) {
      var item = result[i];
      test.ok(item);
      test.ok(typeof item === 'object');

      test.ok(item.exchange);
      test.ok(typeof item.exchange === 'string');

      test.ok(typeof item.priority === 'number');
    }

    test.done();
  });

  checkWrap(test, req);
};


exports.resolveNs = function (test) {
  var req = dns.resolveNs('google.com', function(err, names) {
    test.ifError(err);

    if (names) {
      test.ok(names.length > 0);

      for (var i = 0; i < names.length; i++) {
        var name = names[i];
        test.ok(name);
        test.ok(typeof name === 'string');
      }
    } else {
      test.ok(names);
    }

    test.done();
  });

  checkWrap(test, req);
};


exports.resolveSrv = function (test) {
  var req = dns.resolveSrv('_jabber._tcp.google.com', function(err, result) {
    test.ifError(err);

    test.ok(result);
    test.ok(result.length > 0);

    for (var i = 0; i < result.length; i++) {
      var item = result[i];
      test.ok(item);
      test.ok(typeof item === 'object');

      test.ok(item.name);
      test.ok(typeof item.name === 'string');

      test.ok(typeof item.port === 'number');
      test.ok(typeof item.priority === 'number');
      test.ok(typeof item.weight === 'number');
    }

    test.done();
  });

  checkWrap(test, req);
};


/*
exports.resolveCname = function (test) {
  var req = dns.resolveCname('www.google.com', function(err, names) {
    test.ifError(err);

    test.ok(names);
    test.ok(names.length > 0);

    for (var i = 0; i < names.length; i++) {
      var name = names[i];
      test.ok(name);
      test.ok(typeof name === 'string');
    }

    test.done();
  });

  checkWrap(test, req);
};
*/

exports.resolveTxt = function (test) {
  var req = dns.resolveTxt('backroad.bike', function(err, records) {
    test.ifError(err);
    test.ok(records);
    test.equal(records.length, 1);
    test.equal(records[0][0].indexOf('v=spf1'), 0);
    test.done();
  });

  checkWrap(test, req);
};


exports.lookup_ipv4_explicit = function (test) {
  var req = dns.lookup('www.google.com', 4, function(err, ip, family) {
    test.ifError(err);
    test.ok(net.isIPv4(ip));
    test.strictEqual(family, 4);

    test.done();
  });

  checkWrap(test, req);
};


exports.lookup_ipv4_implicit = function (test) {
  var req = dns.lookup('www.google.com', function(err, ip, family) {
    test.ifError(err);
    test.ok(net.isIPv4(ip));
    test.strictEqual(family, 4);

    test.done();
  });

  checkWrap(test, req);
};


exports.lookup_ipv6_explicit = function (test) {
  var req = dns.lookup('ipv6.google.com', 6, function(err, ip, family) {
    test.ifError(err);
    test.ok(net.isIPv6(ip));
    test.strictEqual(family, 6);

    test.done();
  });

  checkWrap(test, req);
};

/*
// TODO XXX FIXME Doesn't google require ipv6 for this test to pass?
exports.lookup_ipv6_implicit = function (test) {
  var req = dns.lookup('ipv6.google.com', function(err, ip, family) {
    test.ifError(err);
    test.ok(net.isIPv6(ip));
    test.strictEqual(family, 6);

    test.done();
  });

  checkWrap(test, req);
};
//*/

exports.lookup_failure = function (test) {
  var req = dns.lookup('does.not.exist', 4, function(err, ip, family) {
    test.ok(err instanceof Error);
    test.strictEqual(err.errno, dns.NOTFOUND);
    test.strictEqual(err.errno, 'ENOTFOUND');

    test.done();
  });

  checkWrap(test, req);
};

exports.lookup_null = function (test) {
  var req = dns.lookup(null, function(err, ip, family) {
    test.ifError(err);
    test.strictEqual(ip, null);
    test.strictEqual(family, 4);

    test.done();
  });

  checkWrap(test, req);
};


exports.lookup_ip_ipv4 = function (test) {
  var req = dns.lookup('127.0.0.1', function(err, ip, family) {
    test.ifError(err);
    test.strictEqual(ip, '127.0.0.1');
    test.strictEqual(family, 4);

    test.done();
  });

  checkWrap(test, req);
};


//*
exports.lookup_ip_ipv6 = function (test) {
  var req = dns.lookup('::1', function(err, ip, family) {
    test.ifError(err);
    test.ok(net.isIPv6(ip));
    test.strictEqual(family, 6);

    test.done();
  });

  checkWrap(test, req);
};
//*/


exports.lookup_localhost_ipv4 = function (test) {
  var req = dns.lookup('localhost', 4, function(err, ip, family) {
    test.ifError(err);
    test.strictEqual(ip, '127.0.0.1');
    test.strictEqual(family, 4);

    test.done();
  });

  checkWrap(test, req);
};

exports.lookup_longname = function (test) {
  var name, request;

  var name = '************';
  name += name + '***************'
  name += name + '***************'
  name += name + '***************'
  name += name + '***************'
  name += name + '***************'

  request = dns.lookup(name, 4, function (err, ip, family) {
    test.ok(err, "we should fail");
    test.strictEqual(err.errno, dns.NOTFOUND);

    test.done();
  });
  checkWrap(test, request);
};

exports.lookup_without_nameservers = function(test) {
  platform.name_servers = [];
  dns.lookup('www.google.com', function(err) {
    test.ok(err, 'we should fail');
    test.strictEqual(err.errno, dns.TIMEOUT);
    test.done();
    fixupDns();
  });
};

/* Disabled because it appears to be not working on linux. */
/*
exports.lookup_localhost_ipv6 = function (test) {
  var req = dns.lookup('localhost', 6, function(err, ip, family) {
    test.ifError(err);
    test.ok(net.isIPv6(ip));
    test.strictEqual(family, 6);

    test.done();
  });

  checkWrap(test, req);
};
//*/


test/packet.js
var dns = require('../dns'),
  Packet = require('../lib/packet');

exports.roundTrip = function (test) {
  var buff, pre, post;

  pre = new Packet();
  pre.header.id = 12345;
  pre.header.rcode = 1;

  pre.question.push(dns.Question({
    name: 'www.google.com',
    type: dns.consts.NAME_TO_QTYPE.A,
  }));

  pre.answer.push(dns.A({
    name: 'www.google.com',
    address: '127.0.0.1',
    ttl: 600,
  }));

  buff = new Buffer.alloc(1024);

  len = Packet.write(buff, pre);

  post = Packet.parse(buff.slice(0, len));

  test.deepEqual(pre, post);
  test.done();
};



dns.js
// Copyright 2011 Timothy J Fontaine <tjfontaine@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE

"use strict";

exports.platform = require('./lib/platform');

exports.createServer = require('./lib/server').createServer;
exports.createUDPServer = require('./lib/server').createUDPServer;
exports.createTCPServer = require('./lib/server').createTCPServer;

var client = require('./lib/client');
exports.lookup = client.lookup;
exports.resolve = client.resolve;
exports.resolve4 = client.resolve4;
exports.resolve6 = client.resolve6;
exports.resolveMx = client.resolveMx;
exports.resolveTxt = client.resolveTxt;
exports.resolveSrv = client.resolveSrv;
exports.resolveNs = client.resolveNs;
exports.resolveCname = client.resolveCname;
exports.reverse = client.reverse;

var consts = require('native-node-dns-packet').consts;
exports.BADNAME = consts.BADNAME;
exports.BADRESP = consts.BADRESP;
exports.CONNREFUSED = consts.CONNREFUSED;
exports.DESTRUCTION = consts.DESTRUCTION;
exports.REFUSED = consts.REFUSED;
exports.FORMERR = consts.FORMERR;
exports.NODATA = consts.NODATA;
exports.NOMEM = consts.NOMEM;
exports.NOTFOUND = consts.NOTFOUND;
exports.NOTIMP = consts.NOTIMP;
exports.SERVFAIL = consts.SERVFAIL;
exports.TIMEOUT = consts.TIMEOUT;
exports.consts = consts;

var definedTypes = [
  'A',
  'AAAA',
  'NS',
  'CNAME',
  'PTR',
  'NAPTR',
  'TXT',
  'MX',
  'SRV',
  'SOA',
  'TLSA',
].forEach(function (type) {
  exports[type] = function (opts) {
    var obj = {};
    opts = opts || {};
    obj.type = consts.nameToQtype(type);
    obj.class = consts.NAME_TO_QCLASS.IN;
    Object.keys(opts).forEach(function (k) {
      if (opts.hasOwnProperty(k) && ['type', 'class'].indexOf(k) == -1) {
        obj[k] = opts[k];
      }
    });
    return obj;
  };
});

exports.Question = function (opts) {
  var q = {}, qtype;

  opts = opts || {};

  q.name = opts.name;

  qtype = opts.type || consts.NAME_TO_QTYPE.A;
  if (typeof(qtype) === 'string' || qtype instanceof String)
    qtype = consts.nameToQtype(qtype.toUpperCase());

  if (!qtype || typeof(qtype) !== 'number')
    throw new Error("Question type must be defined and be valid");

  q.type = qtype;

  q.class = opts.class || consts.NAME_TO_QCLASS.IN;

  return q;
};
exports.Request = client.Request;


lib/platform.js
// Copyright 2011 Timothy J Fontaine <tjfontaine@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE

'use strict';

var fs = require('fs'),
    EventEmitter = require('events').EventEmitter,
    net = require('net'),
    os = require('os'),
    util = require('util'),
    Cache = require('native-node-dns-cache'),
    consts = require('native-node-dns-packet').consts,
    path = require('path'),
    utils = require('./utils');

var A = consts.NAME_TO_QTYPE.A,
    AAAA = consts.NAME_TO_QTYPE.AAAA,
    PTR = consts.NAME_TO_QTYPE.PTR;

var Platform = function() {
  this._nsReady = false;
  this._hostsReady = false;

  Object.defineProperty(this, 'ready', {
    get: function() {
      return this._nsReady && this._hostsReady;
    }
  });

  this._watches = {};

  Object.defineProperty(this, 'watching', {
    get: function() {
      return Object.keys(this._watches).length > 0;
    },
    set: function(value) {
      var k;
      if (value)
        this._watchFiles();
      else {
        for (k in this._watches) {
          this._watches[k].close();
          delete this._watches[k];
        }
      }
    }
  });

  this.hosts = new Cache();

  this._initNameServers();
  this._initHostsFile();
  this._populate();

  this.cache = false; //new Cache();
};
util.inherits(Platform, EventEmitter);

Platform.prototype.reload = function() {
  this.emit('unready');
  this._initNameServers();
  this._initHostsFile();
  this._populate();
};

Platform.prototype._initNameServers = function() {
  this._nsReady = false;
  this.name_servers = [];
  this.search_path = [];
  this.timeout = 5 * 1000;
  this.attempts = 5;
  this.edns = false;
};

Platform.prototype._initHostsFile = function() {
  this._hostsReady = false;
  this.hosts.purge();
};

Platform.prototype._populate = function() {
  var hostsfile, self = this;

  switch (os.platform()) {
    case 'win32':
      this.name_servers = [
        {
          address: '8.8.8.8',
          port: 53
        },
        {
          address: '8.8.4.4',
          port: 53
        }
      ];
      self._nsReady = true;
      hostsfile = path.join(process.env.SystemRoot,
                        '\\System32\\drivers\\etc\\hosts');
      break;
    default:
      this.parseResolv();
      hostsfile = '/etc/hosts';
      break;
  }

  this._parseHosts(hostsfile);
};

Platform.prototype._watchFiles = function() {
  var self = this, watchParams;

  watchParams = {persistent: false};

  switch (os.platform()) {
    case 'win32':
      //TODO XXX FIXME: it would be nice if this existed
      break;
    default:
      this._watches.resolve = fs.watch('/etc/resolv.conf', watchParams,
          function(event, filename) {
            if (event === 'change') {
              self.emit('unready');
              self._initNameServers();
              self.parseResolv();
            }
          });
      this._watches.hosts = fs.watch('/etc/hosts', watchParams,
          function(event, filename) {
            if (event === 'change') {
              self.emit('unready');
              self._initHostsFile();
              self._parseHosts(hostsfile);
            }
          });
      break;
  }
};

Platform.prototype._checkReady = function() {
  if (this.ready) {
    this.emit('ready');
  }
};

Platform.prototype.parseResolv = function() {
  var self = this;

  fs.readFile('/etc/resolv.conf', 'ascii', function(err, file) {
    if (err) {
      // If the file wasn't found don't worry about it.
      if (err.code == 'ENOENT') {
        return;
      }
      throw err;
    }
    file.split(/\n/).forEach(function(line) {
      var i, parts, subparts;
      line = line.replace(/^\s+|\s+$/g, '');
      if (!line.match(/^#/)) {
        parts = line.split(/\s+/);
        switch (parts[0]) {
          case 'nameserver':
            self.name_servers.push({
              address: parts[1],
              port: 53
            });
            break;
          case 'domain':
            self.search_path = [parts[1]];
            break;
          case 'search':
            self.search_path = [parts.slice(1)];
            break;
          case 'options':
            for (i = 1; i < parts.length; i++) {
              subparts = parts[i].split(/:/);
              switch (subparts[0]) {
                case 'timeout':
                  self.timeout = parseInt(subparts[1], 10) * 1000;
                  break;
                case 'attempts':
                  self.attempts = parseInt(subparts[1], 10);
                  break;
                case 'edns0':
                  self.edns = true;
                  break;
              }
            }
            break;
        }
      }
    });

    self._nsReady = true;
    self._checkReady();
  });
};

Platform.prototype._parseHosts = function(hostsfile) {
  var self = this;

  fs.readFile(hostsfile, 'ascii', function(err, file) {
    var toStore = {};
    if (err) {
      throw err;
    }

    file.split(/\n/).forEach(function(line) {
      var i, parts, ip, revip, kind;
      line = line.replace(/^\s+|\s+$/g, '');
      if (!line.match(/^#/)) {
        parts = line.split(/\s+/);
        ip = parts[0];
        parts = parts.slice(1);
        kind = net.isIP(ip);

        if (parts.length && ip && kind) {
          /* IP -> Domain */
          revip = utils.reverseIP(ip);
          parts.forEach(function(domain) {
            var r = toStore[revip];
            if (!r)
              r = toStore[revip] = {};
            var t = r[PTR];
            if (!t)
              t = r[PTR] = [];
            t.push({
              type: PTR,
              class: 1,
              name: revip,
              data: domain,
              ttl: Infinity
            });
          });

          /* Domain -> IP */
          parts.forEach(function(domain) {
            var r = toStore[domain.toLowerCase()];
            if (!r) {
              r = toStore[domain.toLowerCase()] = {};
            }
            var type = kind === 4 ? A : AAAA;
            var t = r[type];
            if (!t)
              t = r[type] = [];
            t.push({
              type: type,
              name: domain.toLowerCase(),
              address: ip,
              ttl: Infinity
            });
          });
        }
      }
    });

    Object.keys(toStore).forEach(function (key) {
      self.hosts._store.set(self.hosts._zone, key, toStore[key]);
    });
    self._hostsReady = true;
    self._checkReady();
  });
};

module.exports = new Platform();
